tools/gen/impl_template

```diff
@@ -60,25 +60,32 @@ towire_${f.type_obj.name}(${ptr}, ${'' if f.type_obj.is_assignable() else '&'}${
     if f.type_obj.is_varsize():
 	typename += ' *'
 %>\
-% if f.is_varlen():
-${fieldname} = ${f.size('*plen')} ? tal_arr(${ctx}, ${typename}, ${f.size('*plen')}) : NULL;
-% endif
 % if f.is_array() or f.is_varlen():
     % if f.type_obj.has_array_helper():
+    ## We assume array helpers only deal with things literally transcribed!!
+	% if f.is_varlen():
+${fieldname} = ${f.size('*plen')} ? tal_arr(${ctx}, ${typename}, ${f.size('*plen')}) : NULL;
+	% endif
 fromwire_${type_}_array(cursor, plen, ${fieldname}, ${f.size('*plen')});
     % else:
+	% if f.is_varlen():
+${fieldname} = ${f.size('*plen')} ? tal_arr(${ctx}, ${typename}, 0) : NULL;
```

Comment:
the older version loops of these were a bit tighter in that we allocated everything up front instead of in passes, it'd be nice if we could bring that back in a later update.